api/envoy/extensions/transport_sockets/tls/v3/common.proto

```diff
@@ -179,7 +179,6 @@ message PrivateKeyProvider {
     google.protobuf.Any typed_config = 3 [(udpa.annotations.sensitive) = true];
   }
 
-  // [#not-implemented-hide:]
```

Comment:
The fallback won't use the private key material inside each provider, it will use the key at https://github.com/envoyproxy/envoy/blob/74bfcb0b591bb692db0b9b0700b4822ef8a6d777/api/envoy/extensions/transport_sockets/tls/v3/common.proto#L206

In the later, we will have a PR to enable each provider share the same key at 
https://github.com/envoyproxy/envoy/blob/74bfcb0b591bb692db0b9b0700b4822ef8a6d777/api/envoy/extensions/transport_sockets/tls/v3/common.proto#L206 to avoid duplicated key transfer in the sds.

For HSM case, it will ignore the key https://github.com/envoyproxy/envoy/blob/74bfcb0b591bb692db0b9b0700b4822ef8a6d777/api/envoy/extensions/transport_sockets/tls/v3/common.proto#L206

For HSM case about it makes sense to fallback or not(from the security perspective), maybe there are two options:
* Just document clearly that, if you are using the HSM provider, do not fallback. 
 (When using the HSM, people won't fill this field https://github.com/envoyproxy/envoy/blob/74bfcb0b591bb692db0b9b0700b4822ef8a6d777/api/envoy/extensions/transport_sockets/tls/v3/common.proto#L206, then the fallback will failed also)
* Add an interface for each provider to indicate this provider is available for fallback or not. This option probably need us to confirm there is no one want to fallback when using HSM, then we kill this use-case hardly.

Reply:
I'm still not sure it's better to put this here, vs creating a private key provider that itself takes a list of private key providers to try, and defines it's own behavior for what to do if they fail. That delegates the policy decisions to the extension, so it's easier to create new policies later if desired.

Reply:
emm...maybe there are some confuse. @ggreenway correct me if I still misunderstand your point.

Let me explain the whole thing, we already choice the path `creating a private key provider that itself takes a list of private key provider`.

Then the next thing is to implement the fallback, there are two ways:
* Implement the sign/decrypt by the boringSSL method in each provider(or a common provider just for pure boringSSL implement), something like the combine of https://github.com/envoyproxy/envoy/blob/a19966b92bc3afcdd68ebeffe53c4b6848733e79/test/extensions/transport_sockets/tls/test_private_key_method_provider.cc#L35C33-L35C52 and https://github.com/envoyproxy/envoy/blob/a19966b92bc3afcdd68ebeffe53c4b6848733e79/test/extensions/transport_sockets/tls/test_private_key_method_provider.cc#L97
  the downside is we should maintain such number of codes.
* just not set any private key method to the boringSSL method.
  the downside is we touch some core code.

So this PR choice the second way, it much easer and less maintaining burden in the future.

One more solution come up to my mind for `fallback config live in the private key provider itself`, we can add an interface to the PrivateKeyMethodProvider called `bool fallback()` maybe. Then we can move the `fallback` flag into each provider. For provider, they want to implement the fallback, then they can add fallback flag to it config. For the provider never fallback like HSM case, then it can return false always.

Reply:
Got it; I forgot that the default isn't represented (in code) as an extension itself, but just the lack of setting up some callbacks in boringssl. This approach SGTM.